### PR TITLE
Publish a private npm module on PRs, for downstream testing

### PR DESCRIPTION
**Pull request by jessitron**

### CODE REVIEW

File: travis-build.bash [hunk 1]
```diff
@@ -34,28 +34,120 @@ function main () {
...
+            return 1
+          fi
+
+          mv package.json ${temp_package_json}
+          jq ".name=\"${branch_module_name}\"" ${temp_package_json} > package.json
+          if [[ $? -ne 0 || ! -e package.json ]]; then
+            err "failed to update name in package.json"
+            return 1
+          fi
+          rm ${temp_package_json}
```
[the remark this answers is not in the document]
Comment (ddgenome) [reply travis-build.bash]: Also not a bad idea to do this is as a `trap "rm -f $temp_package_json" RETURN` right after it is created so it gets deleted when the function returns. As you have it now, if an error occurs between creation and here, it gets left lying around.

File: travis-build.bash [hunk 2]
```diff
@@ -34,28 +34,120 @@ function main () {
...
+
+          # create this file so we can use npm show on private modules
+           msg "creating local .npmrc using NPM token from environment"
+            if ! ( umask 077 && echo "//registry.npmjs.org/:_authToken=$NPM_TOKEN" > "$HOME/.npmrc" ); then
```
Comment (ddgenome) [inline travis-build.bash]: Indentation is a little odd here.

File: travis-build.bash [hunk 3]
```diff
@@ -34,28 +34,120 @@ function main () {
...
+                err "failed to create $HOME/.npmrc"
                 return 1
             fi
-            local timestamp
-            timestamp=$(date -u +%Y%m%d%H%M%S)
-            if [[ $? -ne 0 || ! $timestamp ]]; then
-                err "failed to generate timestamp"
+            # tell the npm-publish script that we've done this already
+            unset NPM_TOKEN
```
Comment (ddgenome) [inline travis-build.bash]: Why?

File: travis-build.bash [hunk 4]
```diff
@@ -34,28 +34,120 @@ function main () {
...
+
+          # is there already one of these published ?
+          local last_existing_version
+          last_existing_version=$(npm show ${branch_module_name} version)
```
Comment (ddgenome) [inline travis-build.bash]: Braces unnecessary, but you need quotes.

File: travis-build.bash [hunk 5]
```diff
@@ -34,28 +34,120 @@ function main () {
...
+          if [[ $? -ne 0 ]] ; then
```
Comment (ddgenome) [inline travis-build.bash]: Also should test that `last_existing_version` is non-empty.

File: travis-build.bash [hunk 6]
```diff
@@ -34,28 +34,120 @@ function main () {
...
+             # probably not
+             echo "Looks like this is the first time we've published this branch, cool"
+          else
+             # increment the version. First set us to the current, then bump.
+             mv package.json ${temp_package_json}
```
Comment (ddgenome) [inline travis-build.bash]: Check return status. -braces +quotes.

File: travis-build.bash [hunk 7]
```diff
@@ -34,28 +34,120 @@ function main () {
...
+             jq ".version=\"${last_existing_version}\"" ${temp_package_json} > package.json
```
Comment (ddgenome) [inline travis-build.bash]: `jq -e`, check return status, -braces, + quotes

File: travis-build.bash [hunk 8]
```diff
@@ -34,28 +34,120 @@ function main () {
...
+             if [[ $? -ne 0 || ! -e package.json ]]; then
+                err "failed to update version in package.json"
+                return 1
+             fi
+             rm ${temp_package_json}
```
Comment (ddgenome) [inline travis-build.bash]: `trap` would cover this too.

File: travis-build.bash [hunk 9]
```diff
@@ -34,28 +34,120 @@ function main () {
...
+
+             if ! npm version -f patch ; then
+                err "failed to increment version in package.json"
                 return 1
+             fi
+          fi
+
+# what version did we pick?
```
Comment (ddgenome) [inline travis-build.bash]: indent?

File: travis-build.bash [hunk 10]
```diff
@@ -34,28 +34,120 @@ function main () {
...
+          local pkg_version
+          pkg_version=$(jq --raw-output .version package.json)
```
Comment (ddgenome) [inline travis-build.bash]: add `-e`

File: travis-build.bash [hunk 11]
```diff
@@ -34,28 +34,120 @@ function main () {
...
+          if [[ $? -ne 0 || ! $pkg_version ]]; then
+            err "failed to parse version from package.json"
```
Comment (ddgenome) [inline travis-build.bash]: indent

File: travis-build.bash [hunk 12]
```diff
@@ -34,28 +34,120 @@ function main () {
...
+            return 1
+          fi
+
+          if ! bash npm-publish.bash ${pkg_version} restricted ; then
```
Comment (ddgenome) [inline travis-build.bash]: `"$pkg_version"`

With the new `npm-publish.bash` script, this is no longer necessary.

File: travis-build.bash [hunk 13]
```diff
@@ -34,28 +34,120 @@ function main () {
...
+          if ! git checkout package.json ; then
+             echo "WARNING: I changed package.json and couldn't check out the original"
+          fi
+      else
+         echo "No NPM_TOKEN, couldn't publish"
+      fi
+    fi
+
+    if [[ $TRAVIS_PULL_REQUEST == false ]] ; then
+        if [[ $TRAVIS_BRANCH == master || $TRAVIS_TAG =~ ^[0-9]+\.[0-9]+\.[0-9]+(-(m|rc)\.[0-9]+)?$ ]]; then
```
Comment (ddgenome) [inline travis-build.bash]: Seems like these two conditionals could be one?
